bin/phase-helpers.sh

```diff
@@ -843,122 +849,92 @@ __eapi6_src_install() {
 	einstalldocs
 }
 
-# @FUNCTION: has_version
-# @USAGE: [--host-root] <DEPEND ATOM>
-# @DESCRIPTION:
-# Return true if given package is installed. Otherwise return false.
-# Callers may override the ROOT variable in order to match packages from an
-# alternative ROOT.
-has_version() {
-
-	local atom eroot host_root=false root=${ROOT}
-	if [[ $1 == --host-root ]] ; then
-		host_root=true
-		shift
-	fi
+__best_version_and_has_version_common() {
+	local atom root root_arg
+	case $1 in
+		--host-root|-r|-d|-b)
+			root_arg=$1
+			shift ;;
+	esac
 	atom=$1
 	shift
-	[ $# -gt 0 ] && die "${FUNCNAME[0]}: unused argument(s): $*"
+	[ $# -gt 0 ] && die "${FUNCNAME[1]}: unused argument(s): $*"
 
-	if ${host_root} ; then
-		if ! ___eapi_best_version_and_has_version_support_--host-root; then
-			die "${FUNCNAME[0]}: option --host-root is not supported with EAPI ${EAPI}"
-		fi
-		root=/
-	fi
+	case ${root_arg} in
+		"") if ___eapi_has_prefix_variables; then
+				root=${EROOT}
+			else
+				root=${ROOT}
+			fi ;;
+		--host-root)
+			if ! ___eapi_best_version_and_has_version_support_--host-root; then
+				die "${FUNCNAME[1]}: option ${root_arg} is not supported with EAPI ${EAPI}"
+			fi
+			if ___eapi_has_prefix_variables; then
+				root=/${PORTAGE_OVERRIDE_EPREFIX#/}
+			else
+				root=/
+			fi ;;
+		-r|-d|-b)
+			if ! ___eapi_best_version_and_has_version_support_-b_-d_-r; then
+				die "${FUNCNAME[1]}: option ${root_arg} is not supported with EAPI ${EAPI}"
+			fi
+			if ___eapi_has_prefix_variables; then
+				case ${root_arg} in
+					-r) root=${EROOT} ;;
+					-d) root=${ESYSROOT} ;;
```

Comment:
How do we know that `${ESYSROOT}` will be recognized as a valid root inside `pym/portage/package/ebuild/_ipc/QueryCommand.py`, where this code might reject it:
```
		if root not in db:
			return ('', '%s: Invalid ROOT: %s\n' % (cmd, root), 3)
```

Reply:
At this point, it only makes sense for SYSROOT to equal / or ROOT. Until Portage supports a separate third location, setting SYSROOT to anything else is only going to break things anyway. If we accept this then the existing code should be sufficient. I could add explicitly add code to support a different location but I'm struggling to work out what that would look like.

Reply:
Ok, we should die up front if we detect an unsupported SYSROOT configuration. The `profile_check` function inside `pym/_emerge/main.py` would be a reasonable place to report the problem and exit unsuccessfully.

Reply:
I wonder if it will break people's existing SYSROOT usage if emerge dies early though, since I suspect that things like chromium os builds use a SYSROOT variable.

Reply:
@mgorny was trying not to break existing systems so there is a conflict here. On the other hand, it was @vapier who really wanted to called it SYSROOT in the first place.

Reply:
I tried to come up with some code anyway. How about this? I think `settings` is only defining `SYSROOT` in `environ` but it would need to be defined in the constructor for this to work?

```python
if settings['SYSROOT'] != '/' and settings['SYSROOT'] != settings['ROOT']:
	mysettings = config(config_root=config_root, target_root=settings['SYSROOT'],
		env=env, eprefix=eprefix)
	mysettings.lock()
	myroots.append((mysettings['EROOT'], mysettings))
```

Reply:
Well, there aren't any references to SYSROOT in the python code, so maybe adding support in create_trees is overkill. Do we intend for emerge to check install things into SYSROOT, or use packages from SYSROOT to satisify dependencies? If not, we could simply have QueryCommand dynamically instantiate a vardbapi instance in order to handle has_version and best_version calls that refer to SYSROOT.

Reply:
Yes, the final goal would be to install DEPEND into SYSROOT.

Reply:
Wouldn't it be better to let these query commands just blow up if SYSROOT is set to some other location? I mean SYSROOT will only be checked if the -d option is given and this is new in EAPI 7 so it won't break any existing ebuilds. If some new ebuild does use it then the query couldn't possibly return any meaningful answer. We may never implement the additional location (it'd be nice but I don't plan to) and we'll need users to fix their SYSROOT value at some point if we don't want to be in this situation forever.

That's assuming anyone has even set it inappropriately. I'm not familiar with Chromium OS but I've had a look at their wiki. They use the --root-deps option, which implies that they have set SYSROOT to ROOT. They seem to emulate the additional "staging" location by building binary packages of whatever is installed to ROOT and then reinstalling those without build-time dependencies into a final "build_image".

Reply:
Yes, blow up for currently unsupported value is reasonable, but it would be nice if config normalized the value of SYSROOT in order to ensure that QueryCommand recognizes it. For example, the ROOT value is currently normalized in the LocationsManager set_root_override method:
```python
		self.target_root = normalize_path(os.path.abspath(
			self.target_root)).rstrip(os.path.sep) + os.path.sep
```

Reply:
Just thought I'd put in my two cents; I've been testing this locally with the `testpath` script, with
a modified autotools.eclass to use has_version -b in `autotools_env_setup()` and its blowing up
with `has_version: Invalid ROOT: ./`; unsure where the dotslash is coming from following it up
the chain.

Reply:
Thanks for testing, I haven't got very far with mine yet but was about to continue now that I've cleaned things up a bit. I'll certainly look into that.

Reply:
https://gist.github.com/hanetzer/ec9467127f895841c3cd8936288a380c
These are the modifications I've made to the gentoo tree I'm working against.
gentoo/gentoo@96d2e7109eb672f699c50fb2827a29ca1729bb76

Reply:
> Just thought I'd put in my two cents; I've been testing this locally with the testpath script, with
a modified autotools.eclass to use has_version -b in autotools_env_setup() and its blowing up
with has_version: Invalid ROOT: ./; unsure where the dotslash is coming from following it up
the chain.

There's an `Invalid ROOT: ./` fix in 107dca9a3171e4870038769aef66612209ecd423.